stdlib/REPL/src/LineEdit.jl

```diff
@@ -2828,40 +2828,45 @@ function prompt!(term::TextTerminal, prompt::ModalInterface, s::MIState = init_s
     try
         activate(prompt, s, term, term)
         old_state = mode(s)
+        l = Base.ReentrantLock()
+        t = @async while true
+            fcn = take!(s.async_channel)
+            status = @lock l fcn(s)
```

Comment:
I don't think this works. `match_input` needs to be serialized with any potential execution of this function, which may affect the keymap, make arbitrary modifications to the REPL, etc.

Reply:
no, that doesn't work either. The semantics need to be that waiting for the input (through either channel) happens outside the lock, but the modification of the channel `take!` or `read` happens only inside the lock. Otherwise, you're potentially dropping input.